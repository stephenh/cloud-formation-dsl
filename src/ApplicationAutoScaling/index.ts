
export * from './ScalableTarget';
export * from './ScalingPolicy';
