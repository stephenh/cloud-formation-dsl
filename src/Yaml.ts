
/**
 * Yaml-generation helper class.
 *
 * Currently pretty basic, should use something off-the-shelf, but this
 * is cute/easy and might end up getting prettier-type functionality.
 */
export class Yaml {

  output(top: any): string {
    let s = '';
    function concat(prefix: string, o: any) {
      Object.keys(o).forEach(k => {
        const v = o[k];
        if (v instanceof Object) {
          s += `${prefix}${k}:\n`
          concat(prefix + '  ', v);
        } else {
          s += `${prefix}${k}: ${o[k]}\n`;
        }
      });
    }
    concat('', top);
    return s;
  }
}