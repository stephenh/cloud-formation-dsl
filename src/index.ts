
/** Fn::Ref, takes any resources that has a CF name and a marker Ref attribute. */
export function Ref(o: { Name: string, Ref: string }): string {
  return `!Ref: ${o.Name}`
}

/** Shortcut for GetAttr+Arn, takes any resources that has a CF name and a marker Arn attribute. */
export function Arn(o: { Name: string, Arn: string }): string {
  return `!GetAtt: ${o.Name}.${o.Arn}`
}

/** Fn::GetAttr, takes any resources that has a CF name and a marker attribute. */
export function GetAttr<T extends { Name: string }, K extends keyof T>(o: T, k: K): string {
  return `!GetAtt: ${o.Name}.${k}`
}

/** A CloudFormation parameter. */
export function Parameter(name: string) {
  /** Parameters are ref'able, and the value is the name, which is resolved by CF. */
  return { Name: name, Ref: 'Ref' };
}