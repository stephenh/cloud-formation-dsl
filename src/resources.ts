
export const resources: [string, string, any][] = [];

