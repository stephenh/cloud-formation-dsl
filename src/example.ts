
import { AWS_ApplicationAutoScaling_ScalableTarget, AWS_ApplicationAutoScaling_ScalingPolicy } from "@src/ApplicationAutoScaling";
import { resources } from '@src/resources';
import { AWS_IAM_Role } from '@src/IAM';
import { Arn, Parameter, Ref } from '.';

const DynamoMinCapacity = Parameter("DynamoMinCapacity");
const DynamoMaxCapacity = Parameter("DynamoMaxCapacity");
const DynamoDBScaleInOutCooldown = Parameter("DynamoDBScaleInOutCooldown");
const DynamoDBTableScalingThreshold = Parameter("DynamoDBTableScalingThreshold")

const DynamoDBAutoScalingRole = AWS_IAM_Role("DynamoDBAutoScalingRole", {
  AssumeRolePolicyDocument: "..."
});

const EmployeeTableWriteCapacityScalableTarget = AWS_ApplicationAutoScaling_ScalableTarget(
  "EmployeeTableWriteCapacityScalableTarget", {
    ResourceId: "table/Employee",
    MinCapacity: Ref(DynamoMinCapacity),
    MaxCapacity: Ref(DynamoMaxCapacity),
    RoleARN: Arn(DynamoDBAutoScalingRole),
    ScalableDimension: "dynamodb:table:WriteCapacityUnits",
    ServiceNamespace: "dynamodb"
  });

const EmployeeTableWriteScalingPolicy = AWS_ApplicationAutoScaling_ScalingPolicy(
  "EmployeeTableWriteScalingPolicy", {
    PolicyName: "EmployeeTableWriteScalingPolicy",
    PolicyType: "TargetTrackingScaling",
    ScalingTargetId: Ref(EmployeeTableWriteCapacityScalableTarget),
    TargetTrackingScalingPolicyConfiguration: {
      TargetValue: Ref(DynamoDBTableScalingThreshold),
      ScaleInCooldown: Ref(DynamoDBScaleInOutCooldown),
      ScaleOutCooldown: Ref(DynamoDBScaleInOutCooldown),
      PredefinedMetricSpecification: {
        PredefinedMetricType: "DynamoDBWriteCapacityUtilization"
      }
    }
  });

const EmployeeTableReadCapacityScalableTarget = AWS_ApplicationAutoScaling_ScalableTarget(
  "EmployeeTableReadCapacityScalableTarget", {
    ResourceId: "table/Employee",
    MinCapacity: Ref(DynamoMinCapacity),
    MaxCapacity: Ref(DynamoMaxCapacity),
    RoleARN: Arn(DynamoDBAutoScalingRole),
    ScalableDimension: "dynamodb:table:ReadCapacityUnits",
    ServiceNamespace: "dynamodb"
  });

const EmployeeTableReadScalingPolicy = AWS_ApplicationAutoScaling_ScalingPolicy(
  "EmployeeTableReadScalingPolicy", {
    PolicyName: "EmployeeTableReadScalingPolicy",
    PolicyType: "TargetTrackingScaling",
    ScalingTargetId: Ref(EmployeeTableReadCapacityScalableTarget),
    TargetTrackingScalingPolicyConfiguration: {
      TargetValue: Ref(DynamoDBTableScalingThreshold),
      ScaleInCooldown: Ref(DynamoDBScaleInOutCooldown),
      ScaleOutCooldown: Ref(DynamoDBScaleInOutCooldown),
      PredefinedMetricSpecification: {
        PredefinedMetricType: "DynamoDBReadCapacityUtilization"
      }
    }
  });

resources.forEach(resource => {
  const [name, type, props] = resource
  console.log(name)
})
