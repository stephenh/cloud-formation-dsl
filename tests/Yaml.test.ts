
import { Yaml } from '../src/Yaml';

it('should generate basic yaml', () => {
  const o = { key: 'value' };
  const y = new Yaml();
  expect(y.output(o)).toMatchSnapshot();
});

it('should generate two lines', () => {
  const o = { key1: 'value1', key2: 'value2' };
  const y = new Yaml();
  expect(y.output(o)).toMatchSnapshot();
});

it('should generate a child', () => {
  const o = {
    key1: 'value1',
    key2: {
      child1: 'value1'
    }
  };
  const y = new Yaml();
  expect(y.output(o)).toMatchSnapshot();
});