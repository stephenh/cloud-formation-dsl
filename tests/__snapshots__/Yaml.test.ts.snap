// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate a child 1`] = `
"key1: value1
key2:
  child1: value1
"
`;

exports[`should generate basic yaml 1`] = `
"key: value
"
`;

exports[`should generate two lines 1`] = `
"key1: value1
key2: value2
"
`;
